roles/openshift_node/tasks/bootstrap.yml

```diff
@@ -42,6 +45,30 @@
     path: /etc/origin/.config_managed
   register: rpmgenerated_config
 
+- name: create directories for bootstrapping
+  file:
+    state: directory
+    dest: "{{ item }}"
+  with_items:
+  - /root/openshift_bootstrap
+  - /var/lib/origin/openshift.local.config
+  - /var/lib/origin/openshift.local.config/node
+  - "/etc/docker/certs.d/docker-registry.default.svc:5000"
+
+- name: laydown the bootstrap.yml file for on boot configuration
```

Comment:
If the intention here is to deploy ansible playbooks to nodes and execute them later on, I'm against it.  This shouldn't be necessary.  I think what was discussed is we give a bare-bones config to the nodes, and they self configure.  We shouldn't have to do this.

Reply:
@smarterclayton @michaelgugino,
I think there is some deeper discussion that could take place here.  I'm all for building as much into the AMI as possible and having the product do as much as possible.

We have 3 different node groups today with master, infra, and compute.  In our current state, creating a **generic** AMI to become a specific node group without telling it so does not work.  If we want to go down the path of creating an AMI per node group then we can pursue that option but it seems like more work than what we have implemented today.

I think that having the ability to bootstrap gives us more options and more power if needed.  I understand that it is not the only way to accomplish the task but I think that it solves the problem at hand.   @michaelgugino, if you have suggestions please suggest them.

Reply:
If we're going to run ansible against the nodes to configure them, it should be done on the deployment host as it's done today.

I like what @smarterclayton said on the topic, if I understand correctly:
> We should probably consider creating a separate role that can be applied post node setup that uses ansible to ship the bootstrap config into the same location.

Perhaps not a separate role, but an include_role/task_from would be a good fit here.  This is essentially what I was suggesting with my scaleup_ami play, but we'd be pulling the config instead of pushing it.

Reply:
I think the reason @smarterclayton stated this is so that we can bootstrap the normal process and move away from the current implementation in openshift-ansible.  I do not think that he was referring to running openshift-ansible on nodes that have been bootstrapped from an AMI.  If we are going to run openshift-ansible on the nodes post start-up then why bootstrap?  

I think we lose all of the benefits of bootstrap when having to traverse in-and-out of openshift-ansible.  We want less external dependencies and move more configuration into the product.

Reply:
There's two parts of this:

1. nodes that come up from an AMI should never need anything other than the launch config to start (post configuration is an anti-pattern, since it means we have to reintroduce ansible into scale groups)

BUT

2. We want to use bootstrapping on bare metal nodes, which may or may not come up from an image, kickstart, or other flow.  For those, we want a play that can deliver a proper bootstrap config to the standard location on those nodes, exactly like the launch config delivers the bootstrap.

In the second model, the nodes could be configured independently, then masters chosen, and then bootstrap applied.  I suppose there's a third model as well:

3. There's a "pull bootstrapping model" where the kickstart on a bare metal node grabs a kubeconfig from a specific location on demand.

I think the third was what you were referring to?  I'd like to focus on 2 first so we can test bare metal in a controlled fashion, but I don't think 3 is completely out of the picture.

For 1 and 3, ansible wouldn't be involved from a central location.

Reply:
@smarterclayton 

I think what you're proposing is sensible and sounds like a good improvement.  I think we need to draft a spec to formalize these ideas into a series of steps we can implement, especially part 1.  We need to have a formalized spec for the initial launch config, and how we can drive customization there.

Reply:
In regards to my first comment in this thread, about not deploying ansible on the nodes, I still don't prefer it.

However, given the state of the features we wish to implement today, there doesn't seem to be a way around the nodes running some kind of script at boot-time.  That said, that script might as well be in ansible.  I think we should trim as much as we can out of the play we place on the node (see comments in bootstrap.yml), but for now I don't think we should block on this.